util/dvsim/FlowCfg.py

```diff
@@ -470,182 +466,128 @@ def _get_results_page_link(self, relative_to, link_text=''):
+            log.error('Failed to list {} with gsutil. '
+                      'Acting as if there was nothing.'
+                      .format(self.rel_path))
+            existing_paths = []
 
         # Do an ls in the results root dir to check what directories exist.
-        results_dirs = []
-        cmd = self.results_server_cmd + " ls " + self.results_server_path
-        log.log(VERBOSE, cmd)
-        cmd_output = subprocess.run(args=cmd,
-                                    shell=True,
-                                    stdout=subprocess.PIPE,
-                                    stderr=subprocess.DEVNULL)
-        log.log(VERBOSE, cmd_output.stdout.decode("utf-8"))
-        if cmd_output.returncode == 0:
-            # Some directories exist. Check if 'latest' is one of them
-            results_dirs = cmd_output.stdout.decode("utf-8").strip()
-            results_dirs = results_dirs.split("\n")
-        else:
-            log.log(VERBOSE, "Failed to run \"%s\"!", cmd)
-
-        # Start pruning
-        log.log(VERBOSE, "Pruning %s area to limit last 7 results",
-                self.results_server_path)
-
-        rdirs = []
-        for rdir in results_dirs:
-            dirname = rdir.replace(self.results_server_path, '')
-            dirname = dirname.replace('/', '')
-            # Only track history directories with format
-            # "year.month.date_hour.min.sec".
-            if not bool(re.match(r"[\d*.]*_[\d*.]*", dirname)):
-                continue
-            rdirs.append(dirname)
-        rdirs.sort(reverse=True)
-
-        rm_cmd = ""
+        existing_basenames = []
+        for existing_path in existing_paths:
+            # Here, existing_path will start with "gs://" and should end in a
+            # time or with "latest" and then a trailing '/'. Split it to find
+            # that the directory basename. The rsplit() here will result in
+            # ["some_path", "basename_we_want", ""]. Grab the middle.
+            existing_parts = existing_path.rsplit('/', 2)
+            existing_basenames.append(existing_parts[1])
+
+        # We want to add pointers to existing directories with recent
+        # timestamps. Sort in reverse (time and lexicographic!) order, then
+        # take the top few results.
+        existing_basenames.sort(reverse=True)
+
         history_txt = "\n## Past Results\n"
         history_txt += "- [Latest](../latest/" + self.results_html_name + ")\n"
-        if len(rdirs) > 0:
-            for i in range(len(rdirs)):
-                if i < 7:
-                    rdir_url = '../' + rdirs[i] + "/" + self.results_html_name
-                    history_txt += "- [{}]({})\n".format(rdirs[i], rdir_url)
-                elif i > 14:
-                    rm_cmd += self.results_server_path + '/' + rdirs[i] + " "
-
-        if rm_cmd != "":
-            rm_cmd = self.results_server_cmd + " -m rm -r " + rm_cmd + "; "
+        for existing_basename in existing_basenames[:7]:
```

Comment:
Could 7 come out into a constant?

Reply:
I've given it a name locally. Not sure whether it's really worth pulling out further, but it wouldn't be difficult to hoist now.